crates/goose/src/agents/context.rs

```diff
@@ -64,48 +64,62 @@ impl Agent {
-            new_token_counts.insert(0, compaction_marker_tokens);
-
-            // Add an assistant message to continue the conversation
-            let assistant_message = Message::assistant().with_text("
-                The previous message contains a summary that was prepared because a context limit was reached.
-                Do not mention that you read a summary or that conversation summarization occurred
-                Just continue the conversation naturally based on the summarized context
-            ");
-            let assistant_message_tokens: usize = 41;
-            new_messages.push(assistant_message);
-            new_token_counts.push(assistant_message_tokens);
+        // Create the final message list with updated visibility metadata:
+        // 1. Original messages become user_visible but not agent_visible
+        // 2. Summary message becomes agent_visible but not user_visible
+        // 3. Assistant messages to continue the conversation remain both user_visible and agent_visible
+
+        let mut final_messages = Vec::new();
+        let mut final_token_counts = Vec::new();
+
+        // Add all original messages with updated visibility (preserve user_visible, set agent_visible=false)
+        for msg in messages.iter().cloned() {
+            let updated_metadata = msg.metadata.with_agent_invisible();
+            let updated_msg = msg.with_metadata(updated_metadata);
+            final_messages.push(updated_msg);
+            // Token count doesn't matter for agent_visible=false messages, but we'll use 0
+            final_token_counts.push(0);
         }
 
+        // Add the compaction marker (user_visible=true, agent_visible=false)
+        let compaction_marker = Message::assistant()
+            .with_summarization_requested("Conversation compacted and summarized")
+            .with_metadata(MessageMetadata::user_only());
+        let compaction_marker_tokens: usize = 0; // Not counted since agent_visible=false
+        final_messages.push(compaction_marker);
+        final_token_counts.push(compaction_marker_tokens);
+
+        // Add the summary message (agent_visible=true, user_visible=false)
+        let summary_msg = summary_message.with_metadata(MessageMetadata::agent_only());
+        // For token counting purposes, we use the output tokens (the actual summary content)
+        // since that's what will be in the context going forward
+        let summary_tokens = summarization_usage
+            .as_ref()
+            .and_then(|usage| usage.usage.output_tokens)
+            .unwrap_or(0) as usize;
+        final_messages.push(summary_msg);
+        final_token_counts.push(summary_tokens);
+
+        // Add an assistant message to continue the conversation (agent_visible=true, user_visible=false)
+        let assistant_message = Message::assistant()
+            .with_text("The previous message contains a summary that was prepared because a context limit was reached.")
```

Comment:
The 

```
Do not mention that you read a summary or that conversation summarization occurred
Just continue the conversation naturally based on the summarized context
```

seemed helpful to me before in getting it not to make a big deal of the fact that it just read a summary. sometimes it still mentions it, but not nearly as much as if these instructions are removed. given the summary is only visible to the agent, and we want it to feel like we've just continued the conversation I wonder if we can keep this in?

Reply:
Or does this have a better solution now, always inserting the last user message as well?